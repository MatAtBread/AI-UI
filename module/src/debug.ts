// @ts-ignore
export const DEBUG = globalThis.DEBUG == '*' || globalThis.DEBUG == true || globalThis.DEBUG?.match(/(^|\W)AI-UI(\W|$)/) || false;
export function log(...args:any) {
    if (DEBUG) console.log('(AI-UI)', ...args)
}
