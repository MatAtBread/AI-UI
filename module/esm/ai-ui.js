import { isPromiseLike } from './deferred.js';
import { Ignore, asyncIterator, defineIterableProperty, isAsyncIter, isAsyncIterator, iterableHelpers } from './iterators.js';
import { when } from './when.js';
import { UniqueID } from './tags.js';
import { DEBUG, log } from './debug.js';
/* Export useful stuff for users of the bundled code */
export { when } from './when.js';
export * as Iterators from './iterators.js';
let idCount = 0;
const standandTags = [
    "a", "abbr", "address", "area", "article", "aside", "audio", "b", "base", "bdi", "bdo", "blockquote", "body", "br", "button",
    "canvas", "caption", "cite", "code", "col", "colgroup", "data", "datalist", "dd", "del", "details", "dfn", "dialog", "div",
    "dl", "dt", "em", "embed", "fieldset", "figcaption", "figure", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6", "head",
    "header", "hgroup", "hr", "html", "i", "iframe", "img", "input", "ins", "kbd", "label", "legend", "li", "link", "main", "map",
    "mark", "menu", "meta", "meter", "nav", "noscript", "object", "ol", "optgroup", "option", "output", "p", "picture", "pre",
    "progress", "q", "rp", "rt", "ruby", "s", "samp", "script", "search", "section", "select", "slot", "small", "source", "span",
    "strong", "style", "sub", "summary", "sup", "table", "tbody", "td", "template", "textarea", "tfoot", "th", "thead", "time",
    "title", "tr", "track", "u", "ul", "var", "video", "wbr"
];
const elementProtype = {
    get ids() {
        return getElementIdMap(this);
    },
    set ids(v) {
        throw new Error('Cannot set ids on ' + this.valueOf());
    },
    when: function (...what) {
        return when(this, ...what);
    }
};
const poStyleElt = document.createElement("STYLE");
poStyleElt.id = "--ai-ui-extended-tag-styles";
function isChildTag(x) {
    return typeof x === 'string'
        || typeof x === 'number'
        || typeof x === 'function'
        || x instanceof Node
        || x instanceof NodeList
        || x instanceof HTMLCollection
        || x === null
        || x === undefined
        // Can't actually test for the contained type, so we assume it's a ChildTag and let it fail at runtime
        || Array.isArray(x)
        || isPromiseLike(x)
        || isAsyncIter(x)
        || typeof x[Symbol.iterator] === 'function';
}
/* tag */
const callStackSymbol = Symbol('callStack');
export const tag = function (_1, _2, _3) {
    /* Work out which parameter is which. There are 6 variations:
      tag()                                       []
      tag(prototypes)                             [object]
      tag(tags[])                                 [string[]]
      tag(tags[], prototypes)                     [string[], object]
      tag(namespace | null, tags[])               [string | null, string[]]
      tag(namespace | null, tags[], prototypes)   [string | null, string[], object]
    */
    const [nameSpace, tags, prototypes] = (typeof _1 === 'string') || _1 === null
        ? [_1, _2, _3]
        : Array.isArray(_1)
            ? [null, _1, _2]
            : [null, standandTags, _1];
    /* Note: we use property defintion (and not object spread) so getters (like `ids`)
      are not evaluated until called */
    const tagPrototypes = Object.create(null, Object.getOwnPropertyDescriptors(elementProtype));
    // We do this here and not in elementProtype as there's no syntax
    // to copy a getter/setter pair from another object
    Object.defineProperty(tagPrototypes, 'attributes', {
        ...Object.getOwnPropertyDescriptor(Element.prototype, 'attributes'),
        set(a) {
            if (isAsyncIter(a)) {
                const ai = isAsyncIterator(a) ? a : a[Symbol.asyncIterator]();
                const step = () => ai.next().then(({ done, value }) => { assignProps(this, value); done || step(); }, ex => console.warn("(AI-UI)", ex));
                step();
            }
            else
                assignProps(this, a);
        }
    });
    if (prototypes)
        deepDefine(tagPrototypes, prototypes);
    function nodes(...c) {
        const appended = [];
        (function children(c) {
            if (c === undefined || c === null || c === Ignore)
                return;
            if (isPromiseLike(c)) {
                let g = [DomPromiseContainer()];
                appended.push(g[0]);
                c.then(r => {
                    const n = nodes(r);
                    const old = g;
                    if (old[0].parentElement) {
                        appender(old[0].parentElement, old[0])(n);
                        old.forEach(e => e.parentElement?.removeChild(e));
                    }
                    g = n;
                }, (x) => {
                    console.warn('(AI-UI)', x);
                    appender(g[0])(DyamicElementError({ error: x }));
                });
                return;
            }
            if (c instanceof Node) {
                appended.push(c);
                return;
            }
            if (isAsyncIter(c)) {
                const insertionStack = DEBUG ? ('\n' + new Error().stack?.replace(/^Error: /, "Insertion :")) : '';
                const ap = isAsyncIterator(c) ? c : c[Symbol.asyncIterator]();
                // It's possible that this async iterator is a boxed object that also holds a value
                const unboxed = c.valueOf();
                const dpm = (unboxed === undefined || unboxed === c) ? [DomPromiseContainer()] : nodes(unboxed);
                appended.push(...dpm);
                let t = dpm;
                const error = (errorValue) => {
                    const n = t.filter(n => Boolean(n?.parentNode));
                    if (n.length) {
                        t = appender(n[0].parentNode, n[0])(DyamicElementError({ error: errorValue }));
                        n.forEach(e => !t.includes(e) && e.parentNode.removeChild(e));
                    }
                    else
                        console.warn('(AI-UI)', "Can't report error", errorValue, t);
                };
                const update = (es) => {
                    if (!es.done) {
                        try {
                            const n = t.filter(e => e?.parentNode && e.ownerDocument?.body.contains(e));
                            if (!n.length) {
                                // We're done - terminate the source quietly (ie this is not an exception as it's expected, but we're done)
                                throw new Error("Element(s) no longer exist in document" + insertionStack);
                            }
                            t = appender(n[0].parentNode, n[0])(unbox(es.value) ?? DomPromiseContainer());
                            n.forEach(e => !t.includes(e) && e.parentNode.removeChild(e));
                            ap.next().then(update).catch(error);
                        }
                        catch (ex) {
                            // Something went wrong. Terminate the iterator source
                            ap.return?.(ex);
                        }
                    }
                };
                ap.next().then(update).catch(error);
                return;
            }
            if (typeof c === 'object' && c?.[Symbol.iterator]) {
                for (const d of c)
                    children(d);
                return;
            }
            appended.push(document.createTextNode(c.toString()));
        })(c);
        return appended;
    }
    function appender(container, before) {
        if (before === undefined)
            before = null;
        return function (c) {
            const children = nodes(c);
            if (before) {
                // "before", being a node, could be #text node
                if (before instanceof Element) {
                    Element.prototype.before.call(before, ...children);
                }
                else {
                    // We're a text node - work backwards and insert *after* the preceeding Element
                    const parent = before.parentElement;
                    if (!parent)
                        throw new Error("Parent is null");
                    if (parent !== container) {
                        console.warn('(AI-UI)', "Internal error - container mismatch");
                    }
                    for (let i = 0; i < children.length; i++)
                        parent.insertBefore(children[i], before);
                }
            }
            else {
                Element.prototype.append.call(container, ...children);
            }
            return children;
        };
    }
    if (!nameSpace) {
        Object.assign(tag, {
            appender, // Legacy RTA support
            nodes, // Preferred interface instead of `appender`
            UniqueID,
            augmentGlobalAsyncGenerators
        });
    }
    /** Routine to *define* properties on a dest object from a src object **/
    function deepDefine(d, s) {
        if (s === null || s === undefined || typeof s !== 'object' || s === d)
            return;
        for (const [k, srcDesc] of Object.entries(Object.getOwnPropertyDescriptors(s))) {
            try {
                if ('value' in srcDesc) {
                    const value = srcDesc.value;
                    if (value && isAsyncIter(value)) {
                        Object.defineProperty(d, k, srcDesc);
                    }
                    else {
                        // This has a real value, which might be an object, so we'll deepDefine it unless it's a
                        // Promise or a function, in which case we just assign it
                        if (value && typeof value === 'object' && !isPromiseLike(value)) {
                            if (!(k in d)) {
                                // If this is a new value in the destination, just define it to be the same property as the source
                                Object.defineProperty(d, k, srcDesc);
                            }
                            else {
                                if (value instanceof Node) {
                                    log("Having DOM Nodes as properties of other DOM Nodes is a bad idea as it makes the DOM tree into a cyclic graph. You should reference nodes by ID or as a child", k, value);
                                    d[k] = value;
                                }
                                else {
                                    if (d[k] !== value) {
                                        // Note - if we're copying to an array of different length
                                        // we're decoupling common object references, so we need a clean object to
                                        // assign into
                                        if (Array.isArray(d[k]) && d[k].length !== value.length) {
                                            if (value.constructor === Object || value.constructor === Array) {
                                                deepDefine(d[k] = new (value.constructor), value);
                                            }
                                            else {
                                                // This is some sort of constructed object, which we can't clone, so we have to copy by reference
                                                d[k] = value;
                                            }
                                        }
                                        else {
                                            // This is just a regular object, so we deepDefine recursively
                                            deepDefine(d[k], value);
                                        }
                                    }
                                }
                            }
                        }
                        else {
                            // This is just a primitive value, or a Promise
                            if (s[k] !== undefined)
                                d[k] = s[k];
                        }
                    }
                }
                else {
                    // Copy the definition of the getter/setter
                    Object.defineProperty(d, k, srcDesc);
                }
            }
            catch (ex) {
                console.warn('(AI-UI)', "deepAssign", k, s[k], ex);
                throw ex;
            }
        }
    }
    function unbox(a) {
        const v = a?.valueOf();
        return Array.isArray(v) ? v.map(unbox) : v;
    }
    function assignProps(base, props) {
        // Copy prop hierarchy onto the element via the asssignment operator in order to run setters
        if (!(callStackSymbol in props)) {
            (function assign(d, s) {
                if (s === null || s === undefined || typeof s !== 'object')
                    return;
                // static props before getters/setters
                const sourceEntries = Object.entries(Object.getOwnPropertyDescriptors(s));
                sourceEntries.sort((a, b) => {
                    const desc = Object.getOwnPropertyDescriptor(d, a[0]);
                    if (desc) {
                        if ('value' in desc)
                            return -1;
                        if ('set' in desc)
                            return 1;
                        if ('get' in desc)
                            return 0.5;
                    }
                    return 0;
                });
                for (const [k, srcDesc] of sourceEntries) {
                    try {
                        if ('value' in srcDesc) {
                            const value = srcDesc.value;
                            if (isAsyncIter(value)) {
                                assignIterable(value, k);
                            }
                            else if (isPromiseLike(value)) {
                                value.then(value => {
                                    if (value && typeof value === 'object') {
                                        // Special case: this promise resolved to an async iterator
                                        if (isAsyncIter(value)) {
                                            assignIterable(value, k);
                                        }
                                        else {
                                            assignObject(value, k);
                                        }
                                    }
                                    else {
                                        if (s[k] !== undefined)
                                            d[k] = s[k];
                                    }
                                }, error => log("Failed to set attribute", error));
                            }
                            else if (!isAsyncIter(value)) {
                                // This has a real value, which might be an object
                                if (value && typeof value === 'object' && !isPromiseLike(value))
                                    assignObject(value, k);
                                else {
                                    if (s[k] !== undefined)
                                        d[k] = s[k];
                                }
                            }
                        }
                        else {
                            // Copy the definition of the getter/setter
                            Object.defineProperty(d, k, srcDesc);
                        }
                    }
                    catch (ex) {
                        console.warn('(AI-UI)', "assignProps", k, s[k], ex);
                        throw ex;
                    }
                }
                function assignIterable(value, k) {
                    const ap = asyncIterator(value);
                    const update = (es) => {
                        if (!base.ownerDocument.contains(base)) {
                            /* This element has been removed from the doc. Tell the source ap
                              to stop sending us stuff */
                            ap.return?.(new Error("Element no longer exists in document (update " + k.toString() + ")"));
                            return;
                        }
                        if (!es.done) {
                            const value = unbox(es.value);
                            if (typeof value === 'object' && value !== null) {
                                /*
                                THIS IS JUST A HACK: `style` has to be set member by member, eg:
                                  e.style.color = 'blue'        --- works
                                  e.style = { color: 'blue' }   --- doesn't work
                                whereas in general when assigning to property we let the receiver
                                do any work necessary to parse the object. This might be better handled
                                by having a setter for `style` in the PoElementMethods that is sensitive
                                to the type (string|object) being passed so we can just do a straight
                                assignment all the time, or making the decsion based on the location of the
                                property in the prototype chain and assuming anything below "PO" must be
                                a primitive
                                */
                                const destDesc = Object.getOwnPropertyDescriptor(d, k);
                                if (k === 'style' || !destDesc?.set)
                                    assign(d[k], value);
                                else
                                    d[k] = value;
                            }
                            else {
                                // Src is not an object (or is null) - just assign it, unless it's undefined
                                if (value !== undefined)
                                    d[k] = value;
                            }
                            ap.next().then(update).catch(error);
                        }
                    };
                    const error = (errorValue) => {
                        ap.return?.(errorValue);
                        console.warn('(AI-UI)', "Dynamic attribute error", errorValue, k, d, base);
                        appender(base)(DyamicElementError({ error: errorValue }));
                    };
                    ap.next().then(update).catch(error);
                }
                function assignObject(value, k) {
                    if (value instanceof Node) {
                        log("Having DOM Nodes as properties of other DOM Nodes is a bad idea as it makes the DOM tree into a cyclic graph. You should reference nodes by ID or as a child", k, value);
                        d[k] = value;
                    }
                    else {
                        // Note - if we're copying to ourself (or an array of different length),
                        // we're decoupling common object references, so we need a clean object to
                        // assign into
                        if (!(k in d) || d[k] === value || (Array.isArray(d[k]) && d[k].length !== value.length)) {
                            if (value.constructor === Object || value.constructor === Array) {
                                d[k] = new (value.constructor);
                                assign(d[k], value);
                            }
                            else {
                                // This is some sort of constructed object, which we can't clone, so we have to copy by reference
                                d[k] = value;
                            }
                        }
                        else {
                            if (Object.getOwnPropertyDescriptor(d, k)?.set)
                                d[k] = value;
                            else
                                assign(d[k], value);
                        }
                    }
                }
            })(base, props);
        }
    }
    ;
    function tagHasInstance(e) {
        for (let c = e.constructor; c; c = c.super) {
            if (c === this)
                return true;
        }
        return false;
    }
    function extended(_overrides) {
        const instanceDefinition = (typeof _overrides !== 'function')
            ? (instance) => Object.assign({}, _overrides, instance)
            : _overrides;
        const uniqueTagID = Date.now().toString(36) + (idCount++).toString(36) + Math.random().toString(36).slice(2);
        let staticExtensions = instanceDefinition({ [UniqueID]: uniqueTagID });
        /* "Statically" create any styles required by this widget */
        if (staticExtensions.styles) {
            poStyleElt.appendChild(document.createTextNode(staticExtensions.styles + '\n'));
            if (!document.head.contains(poStyleElt)) {
                document.head.appendChild(poStyleElt);
            }
        }
        // "this" is the tag we're being extended from, as it's always called as: `(this).extended`
        // Here's where we actually create the tag, by accumulating all the base attributes and
        // (finally) assigning those specified by the instantiation
        const extendTagFn = (attrs, ...children) => {
            const noAttrs = isChildTag(attrs);
            const newCallStack = [];
            const combinedAttrs = { [callStackSymbol]: (noAttrs ? newCallStack : attrs[callStackSymbol]) ?? newCallStack };
            const e = noAttrs ? this(combinedAttrs, attrs, ...children) : this(combinedAttrs, ...children);
            e.constructor = extendTag;
            const tagDefinition = instanceDefinition({ [UniqueID]: uniqueTagID });
            combinedAttrs[callStackSymbol].push(tagDefinition);
            deepDefine(e, tagDefinition.prototype);
            deepDefine(e, tagDefinition.override);
            deepDefine(e, tagDefinition.declare);
            tagDefinition.iterable && Object.keys(tagDefinition.iterable).forEach(k => {
                if (k in e) {
                    log(`Ignoring attempt to re-define iterable property "${k}" as it could already have consumers`);
                }
                else
                    defineIterableProperty(e, k, tagDefinition.iterable[k]);
            });
            if (combinedAttrs[callStackSymbol] === newCallStack) {
                if (!noAttrs)
                    assignProps(e, attrs);
                for (const base of newCallStack) {
                    const children = base?.constructed?.call(e);
                    if (isChildTag(children)) // technically not necessary, since "void" is going to be undefined in 99.9% of cases.
                        appender(e)(children);
                }
                // Once the full tree of augmented DOM elements has been constructed, fire all the iterable propeerties
                // so the full hierarchy gets to consume the initial state, unless they have been assigned
                // by assignProps from a future
                for (const base of newCallStack) {
                    if (base.iterable)
                        for (const k of Object.keys(base.iterable)) {
                            // We don't self-assign iterables that have themselves been assigned with futures
                            if (!(!noAttrs && k in attrs && (!isPromiseLike(attrs[k]) || !isAsyncIter(attrs[k])))) {
                                const value = e[k];
                                if (value?.valueOf() !== undefined) {
                                    // @ts-ignore - some props of e (HTMLElement) are read-only, and we don't know if k is one of them.
                                    e[k] = value;
                                }
                            }
                        }
                }
            }
            return e;
        };
        const extendTag = Object.assign(extendTagFn, {
            super: this,
            definition: Object.assign(staticExtensions, { [UniqueID]: uniqueTagID }),
            extended,
            valueOf: () => {
                const keys = [...Object.keys(staticExtensions.declare || {}), ...Object.keys(staticExtensions.iterable || {})];
                return `${extendTag.name}: {${keys.join(', ')}}\n \u21AA ${this.valueOf()}`;
            }
        });
        Object.defineProperty(extendTag, Symbol.hasInstance, {
            value: tagHasInstance,
            writable: true,
            configurable: true
        });
        const fullProto = {};
        (function walkProto(creator) {
            if (creator?.super)
                walkProto(creator.super);
            const proto = creator.definition;
            if (proto) {
                deepDefine(fullProto, proto?.prototype);
                deepDefine(fullProto, proto?.override);
                deepDefine(fullProto, proto?.declare);
            }
        })(this);
        deepDefine(fullProto, staticExtensions.prototype);
        deepDefine(fullProto, staticExtensions.override);
        deepDefine(fullProto, staticExtensions.declare);
        Object.defineProperties(extendTag, Object.getOwnPropertyDescriptors(fullProto));
        // Attempt to make up a meaningfu;l name for this extended tag
        const creatorName = fullProto
            && 'className' in fullProto
            && typeof fullProto.className === 'string'
            ? fullProto.className
            : uniqueTagID;
        const callSite = DEBUG ? (new Error().stack?.split('\n')[2] ?? '') : '';
        Object.defineProperty(extendTag, "name", {
            value: "<ai-" + creatorName.replace(/\s+/g, '-') + callSite + ">"
        });
        return extendTag;
    }
    const baseTagCreators = {};
    function createTag(k) {
        if (baseTagCreators[k])
            // @ts-ignore
            return baseTagCreators[k];
        const tagCreator = (attrs, ...children) => {
            let doc = document;
            if (isChildTag(attrs)) {
                children.unshift(attrs);
                attrs = { prototype: {} };
            }
            // This test is always true, but narrows the type of attrs to avoid further errors
            if (!isChildTag(attrs)) {
                if (attrs.debugger) {
                    debugger;
                    delete attrs.debugger;
                }
                if (attrs.document) {
                    doc = attrs.document;
                    delete attrs.document;
                }
                // Create element
                const e = nameSpace
                    ? doc.createElementNS(nameSpace, k.toLowerCase())
                    : doc.createElement(k);
                e.constructor = tagCreator;
                deepDefine(e, tagPrototypes);
                assignProps(e, attrs);
                // Append any children
                appender(e)(children);
                return e;
            }
        };
        const includingExtender = Object.assign(tagCreator, {
            super: () => { throw new Error("Can't invoke native elemenet constructors directly. Use document.createElement()."); },
            extended, // How to extend this (base) tag
            valueOf() { return `TagCreator: <${nameSpace || ''}${nameSpace ? '::' : ''}${k}>`; }
        });
        Object.defineProperty(tagCreator, Symbol.hasInstance, {
            value: tagHasInstance,
            writable: true,
            configurable: true
        });
        Object.defineProperty(tagCreator, "name", { value: '<' + k + '>' });
        // @ts-ignore
        return baseTagCreators[k] = includingExtender;
    }
    tags.forEach(createTag);
    // @ts-ignore
    return baseTagCreators;
};
const { "ai-ui-container": AsyncDOMContainer } = tag('', ["ai-ui-container"]);
const DomPromiseContainer = AsyncDOMContainer.extended({
    styles: `
  ai-ui-container.promise {
    display: ${DEBUG ? 'inline' : 'none'};
    color: #888;
    font-size: 0.75em;
  }
  ai-ui-container.promise:after {
    content: "⋯";
  }`,
    override: {
        className: 'promise'
    },
    constructed() {
        return AsyncDOMContainer({ style: { display: 'none' } }, DEBUG
            ? new Error("Constructed").stack?.replace(/^Error: /, '')
            : undefined);
    }
});
const DyamicElementError = AsyncDOMContainer.extended({
    styles: `
  ai-ui-container.error {
    display: block;
    color: #b33;
    white-space: pre;
  }`,
    override: {
        className: 'error'
    },
    declare: {
        error: undefined
    },
    constructed() {
        if (!this.error)
            return "Error";
        if (this.error instanceof Error)
            return this.error.stack;
        if ('value' in this.error && this.error.value instanceof Error)
            return this.error.value.stack;
        return this.error.toString();
    }
});
export function augmentGlobalAsyncGenerators() {
    let g = (async function* () { })();
    while (g) {
        const desc = Object.getOwnPropertyDescriptor(g, Symbol.asyncIterator);
        if (desc) {
            iterableHelpers(g);
            break;
        }
        g = Object.getPrototypeOf(g);
    }
    if (!g) {
        log("Failed to augment the prototype of `(async function*())()`");
    }
}
export let enableOnRemovedFromDOM = function () {
    enableOnRemovedFromDOM = function () { }; // Only create the observer once
    new MutationObserver(function (mutations) {
        mutations.forEach(function (m) {
            if (m.type === 'childList') {
                m.removedNodes.forEach(removed => removed && removed instanceof Element &&
                    [...removed.getElementsByTagName("*"), removed].filter(elt => !elt.ownerDocument.contains(elt)).forEach(elt => {
                        'onRemovedFromDOM' in elt && typeof elt.onRemovedFromDOM === 'function' && elt.onRemovedFromDOM();
                    }));
            }
        });
    }).observe(document.body, { subtree: true, childList: true });
};
const warned = new Set();
export function getElementIdMap(node, ids) {
    node = node || document;
    ids = ids || {};
    if (node.querySelectorAll) {
        node.querySelectorAll("[id]").forEach(function (elt) {
            if (elt.id) {
                if (!ids[elt.id])
                    ids[elt.id] = elt;
                else if (DEBUG) {
                    if (!warned.has(elt.id)) {
                        warned.add(elt.id);
                        console.info('(AI-UI)', "Shadowed multiple element IDs", elt.id, elt, ids[elt.id]);
                    }
                }
            }
        });
    }
    return ids;
}
