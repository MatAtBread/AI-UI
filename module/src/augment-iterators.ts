import { augmentGlobalAsyncGenerators } from './ai-ui.js';
import { AsyncExtraIterable } from './iterators.js';
export * from './iterators.js';

// Find the prototype that holds the [Symbol.asyncIterator] for an async generator function, and add the helpers to it.
augmentGlobalAsyncGenerators();

export {};

declare global {
  interface AsyncGenerator<T = unknown, TReturn = any, TNext = unknown> extends AsyncIterator<T, TReturn, TNext>, AsyncExtraIterable<T> {
  }
}