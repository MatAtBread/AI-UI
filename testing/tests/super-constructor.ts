/// <reference path="../test.env.d.ts"/>

import { tag, Iterators } from '../../module/src/ai-ui';
const { iterableHelpers } = Iterators;

const sleep = Test.sleep;
let running = true;

const { div, h2, input, span, pre, button } = tag();

const Lazy = h2.extended((instance: { myAttr: number }) => ({
  styles: `
  .Lazy {
    backgroud-color: white;
  }
  `,
  ids: {
    test: span
  },
  declare: {
    Lazy: true,
    myAttr: 57,
    get thing() { return String(instance.myAttr) },
    set thing(v: string) {
      instance.myAttr = Number(v);
      this.dispatchEvent(new Event('change'))
    },
    myMethod(n: number) { instance.myAttr = n },
  },
  override: {
    className: 'Lazy',
    onclick() { this.thing = String(Number(this.thing) - 1) },
    style: {
      borderBottom: '2px solid black'
    }
  },
  async constructed() {
    this.thing = String(this.myAttr);
    await sleep(1);
    return [" Zzz (promise) ", span(this.when('change')(_ => instance.myAttr).initially(instance.myAttr))];
  }
}));

const Lazier = Lazy.extended({
  override: {
    className: `Lazier Lazy`,
    style: {
      borderRight: '2px solid black'
    },
  },
  declare: {
    Lazier: true
  },
  styles: `
    .Lazier {
      font-family: sans-serif;
    }
  `
});

const Laziest = Lazier.extended({
  override: {
    className: `Laziest ${Lazier.className}`,
    onclick() { this.thing = String(Number(this.thing) + 1) },
    style: {
      borderLeft: '2px solid black'
    },
  },
  declare: {
    Laziest: true
  },
  constructed() {
    this.thing = "100";
  },
  styles: `
    .Laziest:hover {
      background-color: #AA0;
    }
  `
});

const App = div.extended({
  styles: `
  .App {
    background-color: #ddd;
    padding: 0.3em;
    position: absolute;
    top: 4px;
    left: 4px;
    bottom: 4px;
    right: 4px;
  }
  `,
  ids: {
    text: input,
    lazy: Laziest
  },
  override: {
    className: 'App'
  },
  constructed() {
    return [
      input({ id: 'text' }),
      this.when('keyup:#text').initially({ start: "Enter some text" }).map(e => ('type' in e) ? div({
        style: {
          color: blinky(this.ids.text!.value)
        }
      }, this.ids.text?.value || e.type) : e.start),
      interval(.25, 10, n => div(n), 'error'),
      interval(.333, 10, n => div(10 - n), 'complete'),
      div({
        style: {
          color: this.when('keyup:#text', blinky('blink'))(e => typeof e === 'string' ? e : this.ids.text!.value)
        },
        onclick() { this.remove() }
      }, 'blink'),
      div({
        style: this.when(blinky('mapped'))(async b => ({ color: b })),
        onclick() { this.remove() }
      }, 'mapped'),
      this.when(blinky('div'))(
        s => div({
          onclick() { this.remove() },
          style: { color: s }
        }, "div")
      ),
      div({ style: { color: 'red' } }, "literal ", 123),
      div({ style: { color: self('red') } }, "generator ", iterableHelpers(self(123)).map(x => " mapped:" + x)),
      div({ style: { color: 'green' } }, "promise ", sleep(1.234, 123)),
      div({ style: { color: this.when(sleep(1, 'red')) } }, "iter/promise ", this.when(sleep(1, 123))),

      div(this.when('#lazy', '@ready')(() => this.ids.lazy!.thing)),

      Laziest({ id: 'lazy' }, 'Lovely!'), pre(Laziest.valueOf()),
      Laziest.super('Super!'), pre(Laziest.super.valueOf()),
      Laziest.super.super('Super duper!'), pre(Laziest.super.super.valueOf()),
      Laziest.super.super.super('Super duper 2'), pre(Laziest.super.super.super.valueOf()),

      this.when('#lazy', '@ready')(() => pre(this.ids.lazy!.constructor.name))
    ]
  }
});

document.body.append(
  App({ style: { border: '1px solid black' } })
);

async function* interval<T>(t: number, limit: number, fn: (n: number) => T, terminate: 'complete' | 'error' = 'complete') {
  for (let i = 0; i < limit; i++) {
    await sleep(t);
    yield fn(i);
  }
  if (terminate === 'error')
    throw new Error("Interval ended");
}

async function* blinky(blinkID: string) {
  console.log("start Blinky", blinkID);
  try {
    while (running) {
      await sleep(.75);
      yield '#808';
      await sleep(.75);
      yield '#088';
    }
  } catch (ex) {
    console.log("Blinky", ex);
  } finally {
    console.log("finally Blinky", blinkID);
  }
  console.log("stop Blinky", blinkID);
}

async function* self<T>(x: T) {
  yield x
}

console.log(document.body.innerHTML);
await sleep(4);
console.log(document.body.innerHTML);
running = false;
await sleep(2);
