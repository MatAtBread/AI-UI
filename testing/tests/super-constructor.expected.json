[
  [1,"start Blinky","blink"],
  [2,"start Blinky","mapped"],
  [3,"start Blinky","div"],
  [4,"<div class=\"App\" style=\"border: 1px solid black;\"><input id=\"text\"><ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container><ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container><ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container><div>blink</div><div>mapped</div><ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container><div style=\"color: red;\">literal 123</div><div>generator <ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></div><div style=\"color: green;\">promise <ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></div><div>iter/promise <ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></div><div><ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></div><h2 class=\"Laziest Lazier Lazy\" style=\"border-bottom: 2px solid black; border-right: 2px solid black; border-left: 2px solid black;\" id=\"lazy\">Lovely!<ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></h2><pre>&lt;ai-Laziest-Lazier-Lazy&gt;: \\{Laziest\\}\n ↪ &lt;ai-Lazier-Lazy&gt;: \\{Lazier\\}\n ↪ &lt;ai-Lazy&gt;: \\{Lazy, myAttr, thing, myMethod\\}\n ↪ TagCreator: &lt;h2&gt;</pre><h2 class=\"Lazier Lazy\" style=\"border-bottom: 2px solid black; border-right: 2px solid black;\">Super!<ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></h2><pre>&lt;ai-Lazier-Lazy&gt;: \\{Lazier\\}\n ↪ &lt;ai-Lazy&gt;: \\{Lazy, myAttr, thing, myMethod\\}\n ↪ TagCreator: &lt;h2&gt;</pre><h2 class=\"Lazy\" style=\"border-bottom: 2px solid black;\">Super duper!<ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></h2><pre>&lt;ai-Lazy&gt;: \\{Lazy, myAttr, thing, myMethod\\}\n ↪ TagCreator: &lt;h2&gt;</pre><h2>Super duper 2</h2><pre>TagCreator: &lt;h2&gt;</pre><ai-ui-container class=\"promise\"><ai-ui-container style=\"display: none;\"></ai-ui-container></ai-ui-container></div>"],
  [5,"<div class=\"App\" style=\"border: 1px solid black;\"><input id=\"text\">Enter some text<ai-ui-container class=\"error\">Error: Interval ended.*</ai-ui-container><div>1</div><div style=\"color: rgb\\(136, 0, 136\\);\">blink</div><div style=\"color: rgb\\(136, 0, 136\\);\">mapped</div><div style=\"color: rgb\\(136, 0, 136\\);\">div</div><div style=\"color: red;\">literal 123</div><div style=\"color: red;\">generator  mapped:123</div><div style=\"color: green;\">promise 123</div><div style=\"color: red;\">iter/promise 123</div><div>100</div><h2 class=\"Laziest Lazier Lazy\" style=\"border-bottom: 2px solid black; border-right: 2px solid black; border-left: 2px solid black;\" id=\"lazy\">Lovely! Zzz \\(promise\\) <span>100</span></h2><pre>&lt;ai-Laziest-Lazier-Lazy&gt;: \\{Laziest\\}\n ↪ &lt;ai-Lazier-Lazy&gt;: \\{Lazier\\}\n ↪ &lt;ai-Lazy&gt;: \\{Lazy, myAttr, thing, myMethod\\}\n ↪ TagCreator: &lt;h2&gt;</pre><h2 class=\"Lazier Lazy\" style=\"border-bottom: 2px solid black; border-right: 2px solid black;\">Super! Zzz \\(promise\\) <span>57</span></h2><pre>&lt;ai-Lazier-Lazy&gt;: \\{Lazier\\}\n ↪ &lt;ai-Lazy&gt;: \\{Lazy, myAttr, thing, myMethod\\}\n ↪ TagCreator: &lt;h2&gt;</pre><h2 class=\"Lazy\" style=\"border-bottom: 2px solid black;\">Super duper! Zzz \\(promise\\) <span>57</span></h2><pre>&lt;ai-Lazy&gt;: \\{Lazy, myAttr, thing, myMethod\\}\n ↪ TagCreator: &lt;h2&gt;</pre><h2>Super duper 2</h2><pre>TagCreator: &lt;h2&gt;</pre><pre>&lt;ai-Laziest-Lazier-Lazy&gt;</pre></div>"],
  [6,"finally Blinky","blink"],
  [7,"stop Blinky","blink"],
  [8,"finally Blinky","mapped"],
  [9,"stop Blinky","mapped"],
  [10,"finally Blinky","div"],
  [11,"stop Blinky","div"]
]